src/lib/ares_process.c

```diff
@@ -816,6 +854,83 @@ static struct server_state *ares__random_server(ares_channel_t *channel)
   return NULL;
 }
 
+/* Pick a server from the list with failover behavior.
+ *
+ * We default to using the first server in the sorted list of servers. That is
+ * the server with the lowest number of consecutive failures and then the
+ * highest priority server (by idx) if there is a draw.
+ *
+ * However, this by itself has two issues:
+ * (1) If a server temporarily goes down and hits some failures, then that
+ *     server will never be retried until all other servers hit the same number
+ *     of failures. This may prevent the server from being retried for a long
+ *     time.
+ * (2) If all servers are failed, then we will continue to select the
+ *     server with the lowest number of consecutive failures even though it may
+ *     be more broken than other servers later in the list.
+ *
+ * For issue (1), with some probability we select a failed server to retry
+ * instead.
+ *
+ * For issue (2), if all servers in the list have a failure, then we fall back
+ * to selecting a random server.
+ */
+static struct server_state *ares__failover_server(ares_channel_t *channel)
+{
+  struct server_state *first_server;
+  struct server_state *last_server;
+  struct timeval       now;
+  unsigned char        c;
+
+  first_server = ares__slist_first_val(channel->servers);
+  last_server  = ares__slist_last_val(channel->servers);
+
+  /* Defensive code against no servers being available on the channel. */
+  if (first_server == NULL) {
+    return NULL;
+  }
+
+  /* If no servers have failures, then prefer the first server in the list. */
+  if (last_server != NULL && last_server->consec_failures == 0) {
+    return first_server;
+  }
+
+  /* If all servers have failures then there is not much reason to prefer one
+   * over another, so select randomly.
+   */
+  if (first_server->consec_failures >= SERVER_ALL_FAILED_THRESHOLD) {
+    return ares__random_server(channel);
+  }
+
+  /* If we are not configured with a server retry chance or the server retry
+   * delay has not yet passed then return the first server in the list.
+   */
+  now = ares__tvnow();
+  if (channel->server_retry_chance == 0 ||
+      !ares__timedout(&now, &channel->server_next_retry_time)) {
+    return first_server;
+  }
+
+  /* Generate a random byte to decide whether to retry a failed server. */
+  ares__rand_bytes(channel->rand_state, &c, 1);
+  if (c % channel->server_retry_chance == 0) {
+    /* Select a failed server to retry. */
+    ares__slist_node_t *node;
+    for (node = ares__slist_node_first(channel->servers); node != NULL;
```

Comment:
probably need to make a helper to get the first failed server (since its a sorted list), then cycle forward from there based on the per-server retry timeout.

Reply:
I think we can do all of this cleanly in a single if statement:
```
      if (node_val != NULL && node_val->consec_failures > 0 &&
          ares__timedout(&now, &node_val->next_retry_time))
```
which skips through all healthy servers, and then skips past ones which haven't passed the retry delay.

Propose that we don't pull this out into a separate helper function.

Reply:
yeah, i guess the likelihood of having an extensive server list is pretty small, likely 2-3 servers.  I was thinking scanning from the end where we know there's a bad server to find the first bad one was more optimal, but that would only be a win with a large number of servers.  I'm fine with this.